src/libcore/iter/mod.rs

```diff
@@ -2242,16 +2242,7 @@ impl<I> Iterator for Take<I> where I: Iterator{
 
     #[inline]
     fn size_hint(&self) -> (usize, Option<usize>) {
-        let (lower, upper) = self.iter.size_hint();
-
-        let lower = cmp::min(lower, self.n);
-
-        let upper = match upper {
-            Some(x) if x < self.n => Some(x),
-            _ => Some(self.n)
-        };
-
-        (lower, upper)
+        TakeImpl::size_hint(self)
```

Comment:
Is the specialization actually needed here?  LLVM already collapses `repeat(x).take(n).size_hint().0` => `n` today ([play demo](https://play.rust-lang.org/?gist=6bffb9248704c9643c9e132ed16e71d0&version=nightly)), since `Repeat::size_hint` is a constant, so I'm not convinced the duplicate size_hint implementation is needed—just the `TrustedLen` marker impls ought to be plenty.

Reply:
The specialization is needed to handle the case of `inner.size_hint()` returning `(x, None)` with `x` being smaller than `self.n`. With the non-specialized implementation this would result in `Take::size_hint` returning `(x, Some(self.n))`, which breaks the contract of `TrustedLen` that the lower and upper bound must be equal if the upper bound is not `None`.

Reply:
Is that allowed?  I thought `TrustedLen` could only have `size_hint`s of `(x, Some(x))` or `(usize::MAX, None)`, since that's the most correct lower bound for "if the actual iterator length is larger than `usize::MAX`".

Reply:
@scottmcm Does the impl here break the contract? Please show an example if that's the case.

Reply:
The trait definition implies that the lower bound is irrelevant ( https://doc.rust-lang.org/std/iter/trait.TrustedLen.html )

> The iterator reports a size hint where it is either exact (lower bound is equal to upper bound), or the upper bound is `None`. The upper bound must only be None if the actual iterator length is larger than `usize::MAX`.

If I recall correctly checks regarding this in libcore are only checking for the lower bound if there is an upper bound.

Reply:
The contract says that the lower and upper bound should be equal, so I think it's important to stick to that (or edit the contract). I can't see any problem in the code in this PR though.

Reply:
I think it would be reasonable to require that implementers of `TrustedLen` return exactly `(usize::MAX, None)` for `size_hint` if there are more than `usize::MAX` elements. If nothing else it would make this specialization unnecessary.

Reply:
@ollie27 While that would make sense and in fact would help regarding https://github.com/rust-lang/rust/pull/47082#issuecomment-361968474 , I don't think that it is relevant for this particular PR. The current guarantees are:

* Either have `Some` upper, in which case the upper and lower bound must be equal,
* Or have `None` as upper.

Those guarantees are reflected in this PR with specialization.

I'll open a thread on internals in a month or so regarding #47082 , where further discussion of these kind of guarantees and optimizations will most likely be done.

Reply:
@bluss I expect the impl is fine; I just am unsure of the necessity.  It feels odd that it would be needed, since the normal logic ought to be fine.

FWIW, if `(0, None)` is a valid `TrustedLen::size_hint`, then `Zip` appears to be wrong, since zipping `(0, None)` and `(10, Some(10))` will give a hint of `(0, 10)`: https://doc.rust-lang.org/nightly/src/core/iter/mod.rs.html#1104-1119 (it's specialized for `TrustedRandomAccess`, but not just `TrustedLen`)

Reply:
I think the correct statement should be "valid if `lower == upper.unwrap_or(usize::MAX)`".

Reply:
@scottmcm Got it, what you're saying makes sense.

Reply:
I think the TrustedLen doc covers this, "An iterator that reports an accurate length using size_hint." It is only accurate if the lower bound is `usize::MAX` if the actual number of elements is usize::MAX or larger.

Reply:
@bluss Where is stated that only a lower bound of `usize::MAX` is correct if there is no upper bound i.e. the iterator is infinite? My understanding was the lower bound can be unspecified because an unbounded iterator does not have a lower bound. It absolutely makes sense to require `usize::MAX` though.
I can't find anything in the documentation of `size_hint` that explicitly states that an unbounded iterator must return `usize::MAX` as lower bound. The only part where it is implied is in the last example:

```rust
// an infinite iterator has no upper bound
// and the maximum possible lower bound
```
I don't know if that's clear enough to make it a trusted guarantee for now. I think it needs more clarification and should be stated explicitly either in `size_hint` or `TrustedLen`. Moreover just to be double sure I think every implementation of `size_hint` on `TrustedLen` iterators should be double-checked to follow this before trusting it.

If I don't miss anything obvious in one of their documentations I'm going to create an issue to check this guarantee, clarify it in the documentation, and remove the specialization from `Take`.

Reply:
@oberien Just IMO here. (0, None) is a "correct" or "true" size_hint for any general iterator (with no further TrustedLen/ExactSizeIterator promises). Because it is a lower bound that does not lie.

If we require that the lower bound is completely exact or "accurate", it must be `n` if the actual length is `n` and it must be `usize::MAX` if the actual length is greater than that.

Reply:
I absolutely agree with you on this one, no discussion there. Infinite iterators / iterators returning more than `usize::MAX` elements implementing TrustedLen should have a defined lower bound.
I'm just saying that this is not (yet) reflected as *explicit* guarantee in the documentation of `size_hint` or `TrustedLen`. But for such memory-safety critical implementation, every guarantee should be as explicit as possible to prevent mistakes.

Reply:
Sure, let's make it more clear

Reply:
I updated the documentation of `TrustedLen` and removed the now unnecessary specialization of `Take<TrustedLen>::size_hint`. Could you please double-check that [all iterators implementing TrustedLen](https://doc.rust-lang.org/nightly/std/iter/trait.TrustedLen.html) actually follow this rule? The only one I found is [`Range<{u,i}32>` if we're on a 16bit platform (i.e. 8bit, which should be negligible)](https://doc.rust-lang.org/nightly/src/core/iter/range.rs.html#168). I don't know if Rust is supporting any 16bit platform.

Reply:
Rust supports 16-bit

Reply:
I opened #48006 for this case.

Reply:
Survey of TrustedLen:

Simple: Empty (0,Some(0)); Once -> Option::IntoIter -> [Item](https://doc.rust-lang.org/nightly/src/core/option.rs.html#1003); the other two Option iterators also go to Item; the result iterators are like the option ones and are just 1 or 0; slice iterators are always finite by definition (they just return (.len(), Some(.len())));  vec IntoIter is also already bounded by usize

Forwards to inner: Rev, Cloned, Map, Enumerate, str::Bytes (to cloned slice iter), &mut I

Interesting: [Chain](https://doc.rust-lang.org/nightly/src/core/iter/mod.rs.html#913-925) makes things longer, and the addition is correct; Zip, as discussed, [already required](https://doc.rust-lang.org/nightly/src/core/iter/mod.rs.html#1104) .0 == .1.unwrap_or(usize::MAX) for correctness

Reply:
Supporting chain was in the design of TrustedLen :smile: